### PR TITLE
change url to hide app behind a relative path

### PR DESCRIPTION
**Pull request by lmazet**
This changes are needed to hide the nodejs service behind a web server with a relative path.

### CODE REVIEW

**Comment by lmazet** [comment]
Exemple of Apache configuration

```
RedirectMatch ^/wbo$ /wbo/

<Proxy http://localhost:3002>
    Allow from all 
</Proxy>

ProxyPass /wbo/ http://localhost:3002/
ProxyPassReverse /wbo/ http://localhost:3002/
ProxyPass /socket.io/ http://localhost:3002/socket.io/
ProxyPassReverse /socket.io/ http://localhost:3002/socket.io/

```

File: server/server.js [hunk 1]
```diff
@@ -83,7 +83,8 @@ function handleRequest(request, response) {
 			// If there is no dot and no directory, parts[1] is the board name
 			boardTemplate.serve(request, response);
 		} else { // Else, it's a resource
-			request.url = "/" + parts.slice(1).join('/');
+			//request.url = "/" + parts.slice(1).join('/');
```
Comment (lovasoa) [inline server/server.js:86]:
```suggestion
```

File: client-data/board.html [hunk 2]
```diff
@@ -5,19 +5,19 @@
...
 		content="{{translations.collaborative_whiteboard}},online,draw,paint,shared,realtime,wbo,whitebophir" />
-	<link rel="apple-touch-icon" href="/favicon.svg">
+	<link rel="apple-touch-icon" href="../favicon.svg">
 	<meta name="viewport" content="width=device-width, user-scalable=no, initial-scale=1.0" />
 	<meta property="og:title" content="{{board}} board on WBO" />
-	<meta property="og:url" content="{{baseUrl}}/boards/{{boardUriComponent}}" />
-	<meta property="og:image" content="{{baseUrl}}/preview/{{boardUriComponent}}" />
-	<link rel="canonical" href="{{baseUrl}}/boards/{{boardUriComponent}}?lang={{language}}" />
+	<meta property="og:url" content="{{boardUriComponent}}" />
+	<meta property="og:image" content="../preview/{{boardUriComponent}}" />
```
Comment (lovasoa) [inline client-data/board.html:17]:
```suggestion
 	<meta property="og:url" content="{{baseUrl}}/boards/{{boardUriComponent}}" />
 	<meta property="og:image" content="{{baseUrl}}/preview/{{boardUriComponent}}" />
```

**Comment by finnboeger** [comment]
I attempted to run your code behind an nginx reverse proxy without success.
You need to specify the path that you would like to run it behind in the configuration and then strip that prefix from the url pathname in `handleRequest` in `server.js`

Static resources are not served properly as well, to correct this you need to wrap change how the `fileserver` variable is declared like so (assuming you set your prefix path in the config as `URL_PREFIX_PATH`):
```js
var fileserver = function(request, response, errorHandleFunction) {
	request.url = request.url.replace(config.URL_PREFIX_PATH, "");
	return serveStatic(config.WEBROOT, {
		maxAge: 2 * 3600 * 1000,
		setHeaders: function (res) {
			res.setHeader("X-UA-Compatible", "IE=Edge");
			res.setHeader("Content-Security-Policy", CSP);
		}
	})(request, response, errorHandleFunction)
};
```
Although I'm sure there is a better way of doing that.

I haven't been able to figure out where socket.js is being served from so far, so that part would still be broken after those changes.

@lovasoa how would you go about serving that file?

**Comment by lovasoa** [comment]
@finnboeger It looks like [the base path can be configured in socket.io too](https://socket.io/docs/server-api/#new-Server-httpServer-options).

We would need something like 

```js
io = iolib(app, {
    path: config.URL_PREFIX_PATH + '/socket.io'
});
```

in https://github.com/lovasoa/whitebophir/blob/master/server/sockets.js#L24

I haven't tried it, though...

**Comment by finnboeger** [comment]
I have implemented the fixes in #57

**Comment by lovasoa** [comment]
Thank you for your contribution @lmazet ! It was merged as a part of #57.

**Comment by lmazet** [comment]
Hi, I agree the contribution's agreement.

**Comment by lmazet** [comment]
By the way, I don't really understand how to setup WBO_ROOT_URL and how to use it. 

I can see behind my proxy :
```
 	<meta property="og:url" content="http://localhost:3002/boards/anonymous" />
 	<meta property="og:image" content="http://localhost:3002/preview/anonymous" />

```but I don't know who or which uses them.

**Comment by finnboeger** [comment]
there is no `WBO_ROOT_URL` configuration setting. I'm assuming you mean `WBO_WEBROOT`?

`WBO_WEBROOT` defines in which directory to look for the static files. Assuming you set up the application by either using the docker image or cloning it and running `npm install && npm start` you won't need to change anything.

How did you access the application? Did you open `localhost:3002` or did you open the url at which your proxy resides? In the latter case it should not display the localhost URL for the opengraph metadata

**Comment by lmazet** [comment]
I know that WBO_ROOT_URL does not exist, @lovasoa suggested me to add it.

I access wbo through my web site `https://www.foobar.com/wbo` from another PC (where localhost is not the wbo server). It works with my Apache configuration settings but not with the one proposed in the  [wiki](https://github.com/lovasoa/whitebophir/wiki/Setup-behind-Reverse-Proxies)
```
RedirectMatch ^/wbo$ /wbo/

<Proxy http://localhost:3002>
        Allow from all 
</Proxy>
ProxyPass /wbo/ http://localhost:3002/
ProxyPassReverse /wbo/ http://localhost:3002/
ProxyPass /socket.io/ http://localhost:3002/socket.io/
ProxyPassReverse /socket.io/ http://localhost:3002/socket.io/

```

**Comment by finnboeger** [comment]
I see, unfortunately I don't have an Apache Server currently set up to test it and am not too familiar with its settings. 

Did you pull the latest master?
And what happens exactly if you run it with your Apache configuration vs what happens with the configuration given in the wiki (changed for your url and port of course)?

**Comment by lovasoa** [comment]
Hello, can you describe your configuration precisely so that we can debug that? 

  - What is the exact commit hash of wbo you are using?
  - How did you launch it? With which environment variables ?
  - What is your exact Apache configuration?
  - What unexpected results do you see? Can you access the index page? Does the websocket connection succeed?

**Comment by lmazet** [comment]
Hi,

Thank for your support and your time. I will try to be more precise:
1) I'm using wbo on commit fce694df (post PR#57 merge).
2) I run wbo on a Debian server and I access it from an Ubuntu machine through an Apache proxy (``https://www.foobar.com/wbo/``) or directly (``http://www:3002/``). I also test from outside my network thought the Apache proxy
3) My Apache configuration is 
```
RedirectMatch ^/wbo$ /wbo/
<Proxy http://localhost:3002>
        Allow from all 
</Proxy>
ProxyPass /wbo/ http://localhost:3002/
ProxyPassReverse /wbo/ http://localhost:3002/
ProxyPass /socket.io/ http://localhost:3002/socket.io/
ProxyPassReverse /socket.io/ http://localhost:3002/socket.io/
```
4) Everything works fine as far as I can see. I can access the index page, In can open a board, I can draw, type text, change colour, zoom, erase, move and change opacity. When I look inside Firefox debugger, I can see that the websocket is working properly. 

My concerns are only:
* The Apache configuration proposed in the [wiki](https://github.com/lovasoa/whitebophir/wiki/Setup-behind-Reverse-Proxies) (i checked that proxy_wstunnel module was activated); I can access the index page but board loading is stuck; the Firefox debugger shown me that the websocket is not open properky.
* When I looked at the source code of a board I can see that the og properties are strangely set:
  - Through ``http://www:3002/``; it looks like ``<meta property="og:url" content="http://www:3002/boards/anonymous" />``
  - Through ``https://www.foobar.com/wbo/``; it looks like ``<meta property="og:url" content="http://localhost:3002/boards/anonymous" />``
  I can't see the impact of this misconfiguration.

Anyway, there is not real issue (expect the fact that the configuration in the [wiki](https://github.com/lovasoa/whitebophir/wiki/Setup-behind-Reverse-Proxies) does not work), just a question about og property settings.

**Comment by finnboeger** [comment]
[og properties](https://ogp.me/) are only used when linking to the page e.g. from facebook or twitter. They enable those boxes with an image, the title and a small description. 
I assume that Apache does not set the `X-Forwarded-Host` header (and probably `X-Forwarded-Proto` as well) correctly. I'm guessing you don't have the `mod_proxy_http` module enabled?

As for the configuration in the wiki @lovasoa might be able to debug it.

**Comment by lmazet** [comment]
Ok. As I don't use Facebook or Twitter, I don't care about this.

No, The module mod_proxy_http is enable on my server. But you were right. There were something strange with my Apache server. I usually test Apache configuration by restarting the daemon (``systemctl restart apache2``). I retried both configurations but this time I stoped and started again the daemon (``systemctl stop apache2; systemctl start apache2``) and both configurations works properly :-) .

So @lovasoa's Apache configuration is ok and I should apology for this long useless thread.

**Comment by lovasoa** [comment]
perfect ! No problem !

**Pull request closed without merging**